File: cpp/assembler.cpp
#include "assembler.hpp"
#include <algorithm>
#include <stdexcept>
#include <iostream>

static std::pair<std::vector<std::string>, size_t> parse_def_args(const std::vector<Token>& tokens, size_t pos) {
    std::vector<std::string> args;
    if (pos < tokens.size() && tokens[pos].type == "LPAREN") {
        pos++;
        while (pos < tokens.size() && tokens[pos].type != "RPAREN") {
            if (tokens[pos].type == "SYMBOL") {
                args.push_back(tokens[pos].value);
            }
            pos++;
            if (pos < tokens.size() && tokens[pos].type == "COMMA") {
                pos++;
            }
        }
        if (pos < tokens.size() && tokens[pos].type == "RPAREN") {
            pos++;
        }
    }
    return {args, pos};
}

static std::pair<std::vector<std::vector<Token>>, size_t> collect_call_args(const std::vector<Token>& tokens, size_t pos, int argc) {
    std::vector<std::vector<Token>> args;
    if (pos < tokens.size() && tokens[pos].type == "LPAREN") {
        pos++;
        for (int a = 0; a < argc; a++) {
            std::vector<Token> arg_tokens;
            while (pos < tokens.size() && tokens[pos].type != "COMMA" && tokens[pos].type != "RPAREN") {
                arg_tokens.push_back(tokens[pos]);
                pos++;
            }
            args.push_back(arg_tokens);
            if (pos < tokens.size() && tokens[pos].type == "COMMA") {
                pos++;
            }
        }
        if (pos < tokens.size() && tokens[pos].type == "RPAREN") {
            pos++;
        }
        return {args, pos};
    }
    return {{}, pos};
}

static std::vector<Token> substitute(const std::vector<Token>& body, const std::vector<std::string>& params, const std::vector<std::vector<Token>>& args) {
    std::vector<Token> result;
    for (const auto& body_token : body) {
        bool substituted = false;
        if (body_token.type == "SYMBOL") {
            for (size_t idx = 0; idx < params.size(); idx++) {
                if (body_token.value == params[idx]) {
                    if (idx < args.size()) {
                        result.insert(result.end(), args[idx].begin(), args[idx].end());
                    }
                    substituted = true;
                    break;
                }
            }
        }
        if (!substituted) {
            result.push_back(body_token);
        }
    }
    return result;
}

static int expand(const std::vector<Token>& tokens, std::map<std::string, Macro>& macros, std::map<std::string, int>& symbols, std::vector<Token>& output, int pc) {
    size_t i = 0;
    while (i < tokens.size()) {
        const auto& t = tokens[i];

        if (t.type == "HASH" && (i + 1) < tokens.size() && tokens[i + 1].value == "define") {
            std::string name = tokens[i + 2].value;
            auto res = parse_def_args(tokens, i + 3);
            std::vector<std::string> params = res.first;
            size_t next_pos = res.second;
            std::vector<Token> body;
            while (next_pos < tokens.size()) {
                const auto& bt = tokens[next_pos];
                // Case insensitive check
                std::string lvalue = bt.value;
                std::transform(lvalue.begin(), lvalue.end(), lvalue.begin(), ::tolower);
                if (bt.type == "SYMBOL" && lvalue == "endm") {
                    next_pos++;
                    break;
                }
                body.push_back(bt);
                next_pos++;
            }
            macros[name] = {params, body, (int)params.size()};
            i = next_pos;
        } else if (t.type == "LABEL") {
            symbols[t.value] = pc;
            i++;
        } else if (t.type == "SYMBOL" && macros.count(t.value)) {
            const auto& macro = macros[t.value];
            auto res = collect_call_args(tokens, i + 1, macro.argc);
            // Collect call args returns empty vector but next_pos might change if () were present.
            // Simplified logic: اگر macro.argc > 0 we expect parens. If argc == 0 parens optional.
            // We'll follow the same pattern as JS/Python.
            // But wait, the JS/Python version returns null for args if LPAREN not found.
            // Let's adjust collect_call_args.
            bool had_parens = (i + 1 < tokens.size() && tokens[i + 1].type == "LPAREN");

            if (had_parens || macro.argc == 0) {
                pc = expand(substitute(macro.body, macro.params, res.first), macros, symbols, output, pc);
                i = res.second;
            } else {
                output.push_back(t);
                pc++;
                i++;
            }
        } else if (t.type == "NUMBER" || t.type == "SYMBOL") {
            output.push_back(t);
            pc++;
            i++;
        } else {
            i++;
        }
    }
    return pc;
}

Pass1Result pass1(const std::string& source) {
    return pass1(tokenize(source));
}

Pass1Result pass1(const std::vector<Token>& tokens) {
    Pass1Result res;
    res.pc = expand(tokens, res.macros, res.symbols, res.tokens, 0);
    return res;
}

std::vector<int> assemble(const std::string& source) {
    Pass1Result res = pass1(source);
    std::vector<int> binary;
    for (const auto& t : res.tokens) {
        if (t.type == "SYMBOL") {
            if (res.symbols.count(t.value)) {
                binary.push_back(res.symbols[t.value]);
            } else {
                throw std::runtime_error("Undefined label: " + t.value);
            }
        } else if (t.type == "NUMBER") {
            binary.push_back(std::stoi(t.value));
        }
    }
    return binary;
}

File: haskell/src/Subleq/Assembler.hs
{-# LANGUAGE OverloadedStrings #-}
module Subleq.Assembler where

import Subleq.Lexer
import Data.Map (Map)
import qualified Data.Map as Map
import Data.Char (toLower)

data MacroDef = MacroDef {
    macroParams :: [String],
    macroBody :: [Token],
    macroArgc :: Int
} deriving (Show, Eq)

data AssemblerState = AssemblerState {
    macros :: Map String MacroDef,
    symbols :: Map String Int,
    pc :: Int,
    tokens :: [Token]
} deriving (Show, Eq)

data Pass1Result = Pass1Result {
    resMacros :: Map String MacroDef,
    resSymbols :: Map String Int,
    resTokens :: [Token],
    resPC :: Int
} deriving (Show, Eq)

parseDefArgs :: [Token] -> Int -> ([String], Int)
parseDefArgs ts pos
    | pos < length ts && tokenType (ts !! pos) == LPAREN = 
        let (args, nextPos) = collectArgs (pos + 1)
        in (args, nextPos)
    | otherwise = ([], pos)
  where
    collectArgs p
        | p >= length ts || tokenType (ts !! p) == RPAREN = ([], p + 1)
        | tokenType (ts !! p) == SYMBOL = 
            let (rest, lastPos) = collectArgs (p + 1)
            in (tokenValue (ts !! p) : rest, lastPos)
        | tokenType (ts !! p) == COMMA = collectArgs (p + 1)
        | otherwise = collectArgs (p + 1)

collectCallArgs :: [Token] -> Int -> Int -> ([[Token]], Int)
collectCallArgs ts pos argc
    | pos < length ts && tokenType (ts !! pos) == LPAREN =
        let (args, nextPos) = collect (pos + 1) 0
        in (args, nextPos)
    | otherwise = ([], pos)
  where
    collect p count
        | count >= argc =
            if p < length ts && tokenType (ts !! p) == RPAREN
            then ([], p + 1)
            else ([], p)
        | p >= length ts = ([], p)
        | otherwise = 
            let (arg, nextP) = collectArg p
                (rest, lastP) = collect nextP (count + 1)
            in (arg : rest, lastP)
    collectArg p
        | p >= length ts || tokenType (ts !! p) == COMMA || tokenType (ts !! p) == RPAREN = ([], p)
        | otherwise = 
            let (rest, nextP) = collectArg (p + 1)
            in (ts !! p : rest, nextP)

substitute :: [Token] -> [String] -> [[Token]] -> [Token]
substitute body params args = concatMap substToken body
  where
    substToken t
        | tokenType t == SYMBOL =
            case Map.lookup (tokenValue t) paramMap of
                Just idx -> args !! idx
                Nothing -> [t]
        | otherwise = [t]
    paramMap = Map.fromList $ zip params [0..]

expand :: [Token] -> Map String MacroDef -> Map String Int -> [Token] -> Int -> (Int, [Token], Map String MacroDef, Map String Int)
expand ts ms ss out c = go 0 c [] ms ss
  where
    go i currPC acc macros symbols
        | i >= length ts = (currPC, reverse acc, macros, symbols)
        | tokenType (ts !! i) == HASH && (i + 1) < length ts && tokenValue (ts !! (i + 1)) == "define" =
            let name = tokenValue (ts !! (i + 2))
                (params, nextPosInitial) = parseDefArgs ts (i + 3)
                (body, nextPos) = collectMacroBody nextPosInitial
                newMacros = Map.insert name (MacroDef params body (length params)) macros
            in go nextPos currPC acc newMacros symbols
        | tokenType (ts !! i) == LABEL =
            let newSymbols = Map.insert (tokenValue (ts !! i)) currPC symbols
            in go (i + 1) currPC acc macros newSymbols
        | tokenType (ts !! i) == SYMBOL && Map.member (tokenValue (ts !! i)) macros =
            let macro = macros Map.! tokenValue (ts !! i)
                (args, nextPos) = collectCallArgs ts (i + 1) (macroArgc macro)
                hasParens = (i + 1 < length ts) && tokenType (ts !! (i + 1)) == LPAREN
            in if hasParens || macroArgc macro == 0
               then let substituted = substitute (macroBody macro) (macroParams macro) args
                        (expandedPC, expandedTokens, newMacros, newSymbols) = expand substituted macros symbols [] currPC
                    in go nextPos expandedPC (reverse expandedTokens ++ acc) newMacros newSymbols
               else go (i + 1) (currPC + 1) (ts !! i : acc) macros symbols
        | tokenType (ts !! i) == SYMBOL || tokenType (ts !! i) == NUMBER =
            go (i + 1) (currPC + 1) (ts !! i : acc) macros symbols
        | otherwise = go (i + 1) currPC acc macros symbols

    collectMacroBody pInitial = 
        let (body, nextP) = goCollect pInitial
        in (body, nextP)
      where
        goCollect p
            | p >= length ts = ([], p)
            | tokenType (ts !! p) == SYMBOL && map toLower (tokenValue (ts !! p)) == "endm" = ([], p + 1)
            | otherwise = 
                let (rest, lastP) = goCollect (p + 1)
                in (ts !! p : rest, lastP)

pass1 :: [Token] -> Pass1Result
pass1 ts = 
    let (finalPC, expandedTokens, finalMacros, finalSymbols) = expand ts Map.empty Map.empty [] 0
    in Pass1Result finalMacros finalSymbols expandedTokens finalPC

assemble :: String -> Either String [Int]
assemble source = do
    case tokenize source of
        Left err -> Left (show err)
        Right ts -> 
            let res = pass1 ts
                results = map (resolve res) (resTokens res)
            in sequence results
  where
    resolve res t
        | tokenType t == SYMBOL =
            case Map.lookup (tokenValue t) (resSymbols res) of
                Just addr -> Right addr
                Nothing -> Left $ "Undefined label: " ++ tokenValue t
        | tokenType t == NUMBER =
            Right $ floor (read (tokenValue t) :: Double)
        | otherwise = Left "Unexpected token during resolution"
